design/007-external-data-handling/004-external-data-handling.md

```diff
@@ -0,0 +1,421 @@
+Description: Data that is very large. Indeed.
+External Sources:
+    - type: s3
+      url: s3://server/path/to/<bucket>
+      location: .renku/mounts/<bucket>
+```
+
+The data is accessible under `data/very-large-data/<bucket>` in the project's directory.
+
+If the data is already mounted by some other means, `dataset stage` can be used
+to let Renku know about it:
+
+```
+renku dataset stage --existing s3://server/path/to/<bucket>:/mounts/bucket
+```
+
+#### Using external data _without_ Renku datasets
+
+We should also consider the case where creating a dataset is too much effort or
+does not fit the stage of the project. In such cases, Renku should allow the
+user to bring in and reference external data. For this, we should review the
+existing `storage` subcommand. For example, to bring in data from an S3 bucket,
+the user would do
+
+```
+renku storage add s3://server/path/to/<bucket>
+-- output --
+<bucket> added as external storage
+```
+
+This would not immediately do anything with the data, but only create the
+`DataStorage` metadata object. For bringing in the data, the user would need
+to do
+
+```
+renku storage stage --copy <bucket> path/to/bucket
+```
+
+The filesystem then looks like
+
+```
+.renku/mounts/<bucket>
+path/to/bucket --> .renku/mounts/<bucket>
+```
+
+The `path/to/bucket` is in `.gitignore`.
+
+The above can be condensed in
+
+```
+renku storage add --copy --to path/to/bucket s3://server/path/to/<bucket>
+```
+
+This is stored in the metadata so when the project is cloned elsewhere, the
+relative location of the data is fixed. So, e.g.
+
+```
+renku storage ls
+-- output --
+Type |             URL              |        Location        |  Project path
+-------------------------------------------------------------------------------
+S3   | s3://server/path/to/<bucket> | .renku/mounts/<bucket> | path/to/<bucket>
+```
+
+To quickly get access to all of the external data when cloning the project
+in a new location, the user can easily mount/copy all the known remote data sources:
+
+```
+renku storage stage --all --copy
+```
+
+Note that in the case of remote storage (as opposed to datasets) there is no
+contradiction in writing directly to the directory where the data source is made
+available in the local project. If the data source is mounted, then the changes
+are reflected on the remote location immediately; if the data is copied, it
+should be possible to implement a `sync` command that would copy upstream new or
+changed files.
+
+#### Streaming external data
+
+The renku-python API should offer some support for more easily accessing remote
+data, especially for sources from which it is common for data to be processed as
+a stream. Ideally, Renku could lazily stage the external data if access to the
+resource is requested.
+
+## Drawbacks
+
+* Relying on FUSE is tricky because it requires elevated privileges on all
+  systems. On Linux, access to /etc/fuse and `CAP_SYS_ADMIN` is needed for fuse to work.
```

Comment:
I believe you mean /dev/fuse ? Also CAP_SYS_ADMIN is only for docker. Or do you mean for setting up fuse?